Ahdevelop

hortonau opened the pull request:
includes my opcode assignments and everyone elses opcodes

hortonau committed:
added austin opcodes

hortonau committed:
fixed some opcode errors

hortonau committed:
added missing semicolon

hortonau committed:
removed some variables that were redefined

hortonau committed:
removed some magic numbers

hortonau committed:
removed more magic nums and printed port

hortonau committed:
added semicolon line 178

hortonau committed:
formatted emulator.c

hortonau committed:
removed magic num on line 85

hortonau committed:
formatting clang

hortonau committed:
Merge remote-tracking branch 'origin/develop' into ah-develop

flatherz commented on emulator.c:
```diff
@@ -39,6 +39,12 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
       }
     case 0x09: // NOLINT
       printf("DAD B");
```
These print statements can be removed. They were just placeholders.

flatherz commented on emulator.c:
```diff
@@ -39,6 +39,12 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
       }
     case 0x09: // NOLINT
       printf("DAD B");
+      uint32_t bc = (cpu->b << 8) | (cpu->c); // NOLINT
```
I think bc and hl can be uint32_t. Result should be uint32_t so that the > 0xFF check works.

flatherz commented on emulator.c:
```diff
@@ -81,6 +87,9 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
     case 0x11: // NOLINT
       printf("LXI D");
+      cpu->d = cpu_read_mem(cpu, cpu->pc + 1);
+      cpu->e = cpu_read_mem(cpu, cpu->pc + 2);
```
I _think_  your registers are reversed here and you need  cpu->e = cpu_read_mem(cpu, cpu->pc + 1) etc due to little endianness.

cliff-bielinski replied:
Yeah e is pc + 1 and d is pc + 2.

flatherz commented on emulator.c:
```diff
@@ -124,6 +133,9 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
       }
     case 0x21: // NOLINT
       printf("LXI H");
+      cpu->h = cpu_read_mem(cpu, cpu->pc + 1);
```
Same comment as above regarding register order and endianness.

flatherz commented on emulator.c:
```diff
@@ -156,6 +168,10 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
       printf("LXI SP");
+      // NOLINTNEXTLINE
+      cpu->sp = cpu_read_mem(cpu, cpu->pc + 1)
+                | (cpu_read_mem(cpu, cpu->pc + 2) << 8); // NOLINT
```
I think this one is correct order though.

flatherz commented on emulator.c:
```diff
@@ -269,6 +289,10 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
+      cpu->c = cpu->memory[cpu->sp];
+      cpu->b = cpu->memory[cpu->sp + 1];
+      cpu->sp += 2;
+      cpu->pc += 1;
```
I don't think you need to increment pc here? 
![image](https://github.com/cliff-bielinski/space-invaders-8080-emulator/assets/122326493/b0d978f7-d17b-48f1-9791-b711b0344991)

cliff-bielinski replied:
Agreed - it's a one opcode instruction.

flatherz commented on emulator.c:
```diff
@@ -302,6 +326,16 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
+      update_carry_flag(cpu, answer > 0xFF); // NOLINT
+      update_aux_carry_flag(cpu, cpu->a, immediate);
+      cpu->a = (uint8_t)answer;
+      cpu->pc += 2;
```
Think you only need to increment pc by 1 here as one is captured by default at the end of the switch statement.

flatherz commented on emulator.c:
```diff
@@ -302,6 +326,16 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
+      update_parity_flag(cpu, (uint8_t)answer);
+      update_carry_flag(cpu, answer > 0xFF); // NOLINT
+      update_aux_carry_flag(cpu, cpu->a, immediate);
+      cpu->a = (uint8_t)answer;
```
I think this works, _might_ be safter to & it with 0x00FF though?

flatherz commented on emulator.c:
```diff
@@ -333,6 +367,10 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
+      // cpu->a will be byte one and port to write to is port 2
+      uint8_t port = cpu_read_mem(cpu, cpu->pc + 1);
+      printf("%u", port);
+      cpu->pc += 2;
```
Same comment as 0xc6 regarding incrementing pc, think you just to need to increment by 1 here.

flatherz commented on emulator.c:
```diff
@@ -358,6 +396,14 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
+      update_parity_flag(cpu, cpu->a);
+      cpu->flags |= FLAG_CY;
+      cpu->flags |= FLAG_AC;
+      cpu->pc += 2;
```
Comment  about incrementing pc here too.

cliff-bielinski replied:
cpu->pc += 1;

flatherz commented on emulator.c:
```diff
@@ -358,6 +396,14 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
+      update_zero_flag(cpu, cpu->a);
+      update_sign_flag(cpu, cpu->a);
+      update_parity_flag(cpu, cpu->a);
+      cpu->flags |= FLAG_CY;
```
Does 0xe6 just clear the flag? I think update_carry_flag() etc is clearer.

hortonau replied:
From the i8080 manual page 4-9
The content of the second byte of the instruction is
logicaJly anded with the contents of the accumu lator .
The result is placed in the accumulator. The CY and
AC flags are cleared.

hortonau replied:
using the update function was clearer. looks like i had AC set to be set instead of be cleared so i updated that as well.

flatherz commented on emulator.c:
```diff
@@ -390,6 +436,8 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
     case 0xfb: // NOLINT
       printf("EI");
+      cpu->interrupt_enabled = true;
+      cpu->pc += 1;
```
Same comment about incrementing pc.

flatherz reviewed (changes requested):
Nice work! You got some interesting instructions to implement. I've got a few things.

- In general, most of the print statements can be removed, as they've been taken out of the other instructions. 
- I also think you started incrementing the pc by one too many partway through. The way I think about it, for most instructions the pc should be incremented by 1 _less than_ the size recorded here: http://www.emulator101.com/reference/8080-by-opcode.html 
- I think the registers are backwards in a few places.
- I see there aren't any unit tests. Are you still working on those? If not, how were you verifying your code?

cliff-bielinski commented on emulator.c:
```diff
@@ -39,6 +39,12 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
+      uint32_t bc = (cpu->b << 8) | (cpu->c); // NOLINT
+      uint32_t hl = (cpu->h << 8) | (cpu->l); // NOLINT
+      uint32_t result = hl + bc;
+      update_carry_flag(cpu, result > 0xFF); // NOLINT
```
I might be wrong, but shouldn't the carry check for result > 0xFFFF? You want to see if the 16-bit value carries over? 0xFF is only 8-bits.

Or if you wanted to maintain 0xFF maybe: (result >> 8) > 0xFF?

hortonau replied:
nope you are not wrong i accidentally kept the same logic i used for checking a carry for adding two 8 bit registers instead of updating for two 16 bit register pairs

cliff-bielinski commented on emulator.c:
```diff
@@ -269,6 +289,10 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
       }
     case 0xc1: // NOLINT
       printf("POP B");
+      cpu->c = cpu->memory[cpu->sp];
```
I would use the function cpu_read_mem rather than directly accessing memory. Brings it in line with the rest of the opcodes and if we decide to change the way the read_mem function works down the road it'll change it here too.

cliff-bielinski commented on emulator.c:
```diff
@@ -358,6 +396,14 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
+      update_zero_flag(cpu, cpu->a);
+      update_sign_flag(cpu, cpu->a);
+      update_parity_flag(cpu, cpu->a);
+      cpu->flags |= FLAG_CY;
```
Isn't this the opposite of what you want? The 8080 manual says this instruction clears the carry flag. I believe |= FLAG_CY would always set the carry flag to 1. Why not just update_carry_flag(cpu, false)?

cliff-bielinski reviewed (changes requested):
Overall nice job but one significant thing is that all these opcodes need unit tests. See comments for other specific changes.

hortonau committed:
made all requested changes 

fixed flipped order for registers
commented out unnecessary prints
updated carry logic to account for 16 bit register pairs

hortonau committed:
fixed formatting issued and implemented requested changes

flatherz commented on .DS_Store:
This can get removed.

flatherz commented on emulator.c:
```diff
@@ -155,8 +172,14 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
     case 0x31: // NOLINT
-      printf("LXI SP");
-      break;
+      // printf("LXI SP");
```
I like the comment indicating which instruction this is, but my personal preference is that you remove the printf part call -- makes it look like commented out code instead of an intentional comment!

flatherz commented on emulator.c:
```diff
@@ -268,8 +301,14 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
+        cpu->c = cpu_read_mem(cpu, cpu->sp);
+        cpu->b = cpu_read_mem(cpu, cpu->sp + 1);
+        cpu->sp += 2;
+        // NOT NEEDED cpu->pc += 1;
```
Please remove commented out code.

flatherz commented on emulator.o:
Please remove this too!

flatherz reviewed (changes requested):
Just a few cosmetic things to update on this code.

hortonau committed:
added unit tests for austin opcodes

hortonau committed:
added nolint as temporary workaround

hortonau committed:
NOLINT case sensitive

hortonau committed:
formatted tests.c file

flatherz commented on tests.c:
```diff
@@ -63,6 +74,33 @@ test_opcode_0x06(void) // NOLINT
+  CU_ASSERT(cpu.pc == initial_pc + 1);
+
+  CU_ASSERT(result_hl == initial_hl + initial_bc);
+  CU_ASSERT((cpu.flags & FLAG_CY) == (carry_occurred ? FLAG_CY : 0));
```
I think there should  probably be explicit tests of a case where carry occurs _and_ a case where it doesn't. If I understand this line correctly, it would account for both cases, but the test is only run for one. Won't require changing it for the review though.

cliff-bielinski commented on tests.c:
```diff
@@ -63,6 +74,33 @@ test_opcode_0x06(void) // NOLINT
+
+  CU_ASSERT(result_hl == initial_hl + initial_bc);
+  CU_ASSERT((cpu.flags & FLAG_CY) == (carry_occurred ? FLAG_CY : 0));
+  cpu.b = 0;
+  cpu.c = 0;
+  cpu.h = 0;
+  cpu.l = 0;
```
Setting registers back to 0 is probably unnecessary since cpu_init handles that. No need to change though.
